pkg/service/mempool.go

```diff
@@ -50,36 +51,58 @@ type flushSpendReq struct {
 
 func NewMempool(storage db.Storage, client eth.Client) *Mempool {
 	return &Mempool{
-		txReqs:     make(chan *txRequest),
-		quit:       make(chan bool),
-		flushReq:   make(chan flushSpendReq),
-		txPool:     make([]MempoolTx, 0),
-		poolSpends: make(map[string]bool),
-		storage:    storage,
-		client:     client,
+		txReqs:           make(chan *txRequest),
+                validatedTxReqs:  make(chan *txRequest),
+		quit:             make(chan bool),
+		flushReq:         make(chan flushSpendReq),
+		txPool:           make([]MempoolTx, 0),
+		poolSpends:       make(map[string]bool),
+		storage:          storage,
+		client:           client,
 	}
 }
 
+func (m *Mempool) validateTransactionRequest(req *txRequest) {
+        tx := req.tx
+        var err error
+        if tx.Body.IsDeposit() {
+                err = m.VerifyDepositTransaction(&tx)
+        } else {
+                // 300-500us
+                err = m.VerifySpendTransaction(&tx)
+        }
+        if err != nil {
+                mPoolLogger.WithFields(logrus.Fields{
+                        "hash":   tx.Body.SignatureHash().Hex(),
+                        "reason": err,
+                }).Warn("transaction rejected from mempool")
+
+                req.res <- TxInclusionResponse{
+                        Error: err,
+                }
+                return
+        }
+
+        m.validatedTxReqs <- req
+}
+
 func (m *Mempool) Start() error {
 	go func() {
 		for {
 			select {
 			case req := <-m.txReqs:
-				if len(m.txPool) == MaxMempoolSize {
-					req.res <- TxInclusionResponse{
-						Error: errors.New("mempool is full"),
-					}
-					continue
-				}
-
-				tx := req.tx
-				var err error
-				if tx.Body.IsDeposit() {
-					err = m.VerifyDepositTransaction(&tx)
-				} else {
-					err = m.VerifySpendTransaction(&tx)
-				}
-				if err != nil {
+                                go m.validateTransactionRequest(req)
+                        case req := <-m.validatedTxReqs:
```

Comment:
We validate the incoming transaction request first against previous block history for double-spends and once that's cleared we do our mem-pool integrity checks synchronously below.

Reply:
I like this approach.